File: CpuA64/assembler.py
#!/usr/bin/python3

"""
This files contains ELF like abstraction to help build an a64 assembler.
"""
from typing import List, Dict, Optional, Any

import CpuA64.opcode_tab as a64
from CpuA64 import disass
import Elf.elfhelper as elf
import Elf.enum_tab as elf_enum
from Util import parse

ZERO_BYTE = bytes([0])
NOP_BYTES = bytes([0x1f, 0x20, 0x03, 0xd5])


def DumpData(data: bytes, addr: int, syms: Dict[int, Any]) -> str:
    out = []
    first_address = True
    for n, b in enumerate(data):
        if first_address or (addr + n) % 8 == 0 or (addr + n) in syms:
            out.append(f"{addr:06x}")
            first_address = False
            name = syms.get(addr + n)
            if name:
                out[-1] += f" [{name}]"
        out[-1] += f" {b:02x}"
    return "\n".join(out)


class Unit:
    """Holds a low level binary version of an A32 executable """

    def __init__(self):
        self.sec_text = elf.Section.MakeSectionText(1)
        self.sec_rodata = elf.Section.MakeSectionRodata(1)
        self.sec_data = elf.Section.MakeSectionData(1)
        self.sec_bss = elf.Section.MakeSectionBss(1)
        #
        self.global_symbol_map: Dict[str, elf.Symbol] = {}
        self.symbols: List[elf.Symbol] = []
        self.relocations: List[elf.Reloc] = []

        # used while processing
        self.local_symbol_map: Dict[str, elf.Symbol] = {}
        self.mem_sec: Optional[elf.Section] = None
        self.current_fun: Optional[str] = None

    def AddSymbol(self, name, sec: Optional[elf.Section], is_local: bool) -> elf.Symbol:
        the_map = self.local_symbol_map if is_local else self.global_symbol_map
        sym = the_map.get(name)
        if sym is None:
            # ~0 is our undefined symbol marker. It is checked in
            val = ~0 if sec is None else len(sec.data)
            sym = elf.Symbol.Init(name, is_local, sec, val)
            self.symbols.append(sym)
            the_map[name] = sym
            return sym
        elif sec is not None:
            # the symbol was forward declared and now we are filling in the missing info
            assert sym.is_undefined(), f"{sym} already defined"
            sym.section = sec
            sym.st_value = len(sec.data)
        return sym

    def FindOrAddSymbol(self, name, is_local) -> elf.Symbol:
        the_map = self.local_symbol_map if is_local else self.global_symbol_map

        sym = the_map.get(name)
        if sym is None:
            return self.AddSymbol(name, None, is_local)
        return sym

    def AddReloc(self, reloc_kind, sec: elf.Section, symbol: elf.Symbol,
                 extra: int):
        self.relocations.append(
            elf.Reloc.Init(reloc_kind.value, sec, len(sec.data), symbol, extra))

    def FunStart(self, name: str, alignment: int):
        self.sec_text.PadData(alignment, NOP_BYTES)
        self.AddSymbol(name, self.sec_text, False)
        assert self.current_fun is None
        self.current_fun = name

    def FunEnd(self):
        assert self.current_fun is not None
        self.current_fun = None
        self.local_symbol_map.clear()

    def MemStart(self, name: str, alignment: int, kind: str, is_local_sym):
        assert self.mem_sec is None
        if kind == "rodata":
            self.mem_sec = self.sec_rodata
        elif kind == "data":
            self.mem_sec = self.sec_data
        elif kind == "bss":
            self.mem_sec = self.sec_bss
        else:
            assert False, f"bad mem kind {kind}"
        self.mem_sec.PadData(alignment, ZERO_BYTE)
        self.AddSymbol(name, self.mem_sec, is_local_sym)

    def MemEnd(self):
        assert self.mem_sec is not None
        self.mem_sec = None

    def AddData(self, repeats: int, data: bytes):
        assert self.mem_sec is not None
        self.mem_sec.AddData(data * repeats)

    def AddFunAddr(self, reloc_type, size: int, fun_name: str):
        assert size == 4
        assert self.mem_sec is not None
        symbol = self.FindOrAddSymbol(fun_name, False)
        self.AddReloc(reloc_type, self.mem_sec, symbol, 0)
        self.mem_sec.AddData(b"\0" * size)

    def AddBblAddr(self, reloc_type, size: int, bbl_name: str):
        assert size == 4
        assert self.current_fun is not None
        assert self.mem_sec is not None
        symbol = self.FindOrAddSymbol(bbl_name, True)
        self.AddReloc(reloc_type, self.mem_sec, symbol, 0)
        self.mem_sec.AddData(b"\0" * size)

    def AddMemAddr(self, reloc_type, size: int, mem_name: str, addend: int):
        assert size == 4
        assert self.mem_sec is not None
        symbol = self.FindOrAddSymbol(mem_name, False)
        self.AddReloc(reloc_type, self.mem_sec, symbol, addend)
        self.mem_sec.AddData(b"\0" * size)

    def AddLabel(self, name: str, alignment: int, padding_bytes: bytes):
        assert alignment % len(padding_bytes) == 0
        self.sec_text.PadData(alignment, padding_bytes)
        assert self.current_fun is not None
        self.AddSymbol(name, self.sec_text, True)

    def AddIns(self, ins: a64.Ins):
        if ins.reloc_kind != elf_enum.RELOC_TYPE_AARCH64.NONE:
            sym = self.FindOrAddSymbol(ins.reloc_symbol, ins.is_local_sym)
            self.AddReloc(ins.reloc_kind, self.sec_text, sym, ins.operands[ins.reloc_pos])
            # clear reloc info before proceeding
            ins.reloc_kind = elf_enum.RELOC_TYPE_AARCH64.NONE
            ins.operands[ins.reloc_pos] = 0
        self.sec_text.AddData(a64.Assemble(ins).to_bytes(4, byteorder='little'))

    def AddLinkerDefs(self):
        """must be called last"""
        if self.sec_bss.sh_size > 0:
            self.sec_bss.PadData(16, ZERO_BYTE)
            self.AddSymbol("$$rw_data_end", self.sec_bss, False)
        elif self.sec_data.sh_size > 0:
            self.sec_data.PadData(16, ZERO_BYTE)
            self.AddSymbol("$$rw_data_end", self.sec_data, False)

    def AddStartUpCode(self):
        """Add code for `_start` wrapper which calls main(()

        When Linux transfers control to a new A32 program is does not follow any calling
        convention so we need this shim.
        The initial execution env looks like this:
        0(sp)			argc

        4(sp)			    argv[0] # argv start
        8(sp)               argv[1]
        ...
        (4*argc)(sp)        NULL    # argv sentinel

        (4*(argc+1))(sp)    envp[0] # envp start
        (4*(argc+2))(sp)    envp[1]
        ...
                            NULL    # envp sentinel
        """
        assert False
        self.FunStart("_start", 16)
        for mnemonic, ops in [
            ("ldr_imm_add", "al r0 sp 0"),
            ("add_imm", "al r1 sp 4"),
            ("bl", "al expr:call:main"),
            ("movw", "al r7 1"),
            ("svc", "al 0"),
            ("ud2", "al")]:
            self.AddIns(dis.InsParse(mnemonic, ops.split()))
        self.FunEnd()

    def __str__(self):
        syms = {}
        return f"""UNIT
SECTION[text] {len(self.sec_text.data)}
{DumpData(self.sec_text.data, 0, syms)}
SECTION[rodata] {len(self.sec_rodata.data)}
{DumpData(self.sec_rodata.data, 0, syms)}     
SECTION[data] {len(self.sec_data.data)}
{DumpData(self.sec_data.data, 0, syms)}    
SECTION[bss] {len(self.sec_bss.data)}
{DumpData(self.sec_bss.data, 0, syms)}    
"""


def HandleOpcode(mnemonic, token: List[str], unit: Unit):
    ins = disass.InsParse(mnemonic, token)
    if ins.reloc_kind != elf_enum.RELOC_TYPE_AARCH64.NONE:
        sym = unit.FindOrAddSymbol(ins.reloc_symbol, ins.is_local_sym)
        unit.AddReloc(ins.reloc_kind, unit.sec_text, sym, ins.operands[ins.reloc_pos])
        # clear reloc info before proceeding
        ins.reloc_kind = elf_enum.RELOC_TYPE_AARCH64.NONE
        ins.operands[ins.reloc_pos] = 0
    unit.sec_text.AddData(a64.Assemble(ins).to_bytes(4, byteorder='little'))



class ParseError(Exception):
    pass


def UnitParse(fin, add_startup_code) -> Unit:
    unit = Unit()
    dir_handlers = {
        ".fun": lambda x, y: unit.FunStart(x, int(y, 0)),
        ".endfun": unit.FunEnd,
        ".mem": lambda x, y, z: unit.MemStart(x, int(y, 0), z, False),
        ".localmem": lambda x, y, z: unit.MemStart(x, int(y, 0), z, True),
        ".endmem": unit.MemEnd,
        ".data": lambda x, y: unit.AddData(int(x, 0),
                                           parse.QuotedEscapedStringToBytes(y)),
        ".addr.fun": lambda x, y: unit.AddFunAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y),
        ".addr.bbl": lambda x, y: unit.AddBblAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y),
        ".addr.mem": lambda x, y, z: unit.AddMemAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y, int(z, 0)),
        ".bbl": lambda x, y: unit.AddLabel(x, int(y, 0), NOP_BYTES),
    }
    for line_num, line in enumerate(fin):
        token = parse.ParseLine(line)
        # print(line)
        if not token:
            continue
        mnemonic = token.pop(0)

        if mnemonic.startswith("#"):
            continue
        if mnemonic.startswith("."):
            handler = dir_handlers.get(mnemonic)
            assert handler is not None
            handler(*token)
            continue

        try:
            HandleOpcode(mnemonic, token, unit)
        except Exception as err:
            raise ParseError(
                f"UnitParseFromAsm error in line {line_num}:\n{line}\n{token}\n{err}")
    unit.AddLinkerDefs()
    if add_startup_code:
        unit.AddStartUpCode()
    return unit


# sample 90000440
_OPCODE_ADRP: a64.Opcode = a64.Opcode.name_to_opcode["adrp"]
# sample 9126c000
_OPCODE_ADD_X_IMM: a64.Opcode = a64.Opcode.name_to_opcode["add_x_imm"]
# sample 14014192
_OPCODE_B: a64.Opcode = a64.Opcode.name_to_opcode["b"]
# sample 97fffcf7
_OPCODE_BL: a64.Opcode = a64.Opcode.name_to_opcode["bl"]
# sample 54000140
_OPCODE_COND_BR = [a64.Opcode.name_to_opcode[f"b_{cond}"] for cond in a64.CONDITION_CODES]


def _patch_ins(ins_old: int, opcode: a64.Opcode, pos: int, value: int):
    ops = opcode.DisassembleOperands(ins_old)
    ops[pos] = value
    return opcode.AssembleOperands(ops)


def _branch_offset(rel: elf.Reloc, sym_val: int) -> int:
    return (sym_val - (rel.section.sh_addr + rel.r_offset)) >> 2


def _adrp_offset(rel: elf.Reloc, sym_val: int) -> int:
    return (sym_val >> 12) - ((rel.section.sh_addr + rel.r_offset) >> 12)


def _ApplyRelocation(rel: elf.Reloc):
    sec_data = rel.section.data
    sym_val = rel.symbol.st_value + rel.r_addend
    assert rel.r_offset + 4 <= len(sec_data)
    old_data = int.from_bytes(sec_data[rel.r_offset:rel.r_offset + 4], "little")

    if rel.r_type == elf_enum.RELOC_TYPE_AARCH64.ADR_PREL_PG_HI21.value:
        new_data = _patch_ins(old_data, _OPCODE_ADRP, 1, _adrp_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC.value:
        new_data = _patch_ins(old_data, _OPCODE_ADD_X_IMM, 2, sym_val & 0xfff)
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.CONDBR19.value:
        new_data = _patch_ins(old_data, _OPCODE_COND_BR[old_data & 0xf], 0, _branch_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.JUMP26.value:
        new_data = _patch_ins(old_data, _OPCODE_B, 0, _branch_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.CALL26.value:
        new_data = _patch_ins(old_data, _OPCODE_BL, 1, _branch_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.ABS32.value:
        new_data = sym_val
    elif rel.r_type == elf_enum.RELOC_TYPE_AARCH64.ABS64.value:
        new_data = sym_val
    else:
        assert False, f"unknown kind reloc {rel}"

    sec_data[rel.r_offset:rel.r_offset + 4] = new_data.to_bytes(4, "little")
    print(f"PATCH INS {rel.r_type} {rel.r_offset:x} {sym_val:x} {old_data:x} {new_data:x} {rel.symbol.name}")


def Assemble(unit: Unit, create_sym_tab: bool) -> elf.Executable:
    sections = []
    segments = []

    seg_exe = elf.Segment.MakeExeSegment(65536)
    segments.append(seg_exe)

    sec_null = elf.Section.MakeSectionNull()
    sections.append(sec_null)
    seg_exe.sections.append(sec_null)
    #
    sec_text = unit.sec_text
    assert len(sec_text.data) > 0
    sections.append(sec_text)
    seg_exe.sections.append(sec_text)

    sec_rodata = unit.sec_rodata
    if len(sec_rodata.data) > 0:
        seg_ro = elf.Segment.MakeROSegment(65536)
        segments.append(seg_ro)
        #
        sections.append(sec_rodata)
        seg_ro.sections.append(sec_rodata)

    if len(unit.sec_data.data) + len(unit.sec_bss.data) > 0:
        seg_rw = elf.Segment.MakeRWSegment(65536)
        segments.append(seg_rw)

    sec_data = unit.sec_data
    if len(sec_data.data) > 0:
        sections.append(sec_data)
        seg_rw.sections.append(sec_data)

    sec_bss = unit.sec_bss
    if len(sec_bss.data) > 0:
        sections.append(sec_bss)
        seg_rw.sections.append(sec_bss)

    seg_pseudo = elf.Segment.MakePseudoSegment()
    segments.append(seg_pseudo)
    #

    if create_sym_tab:
        # we do not create the content here since we cannot really do this until
        # the section addresses are finalized
        which = elf_enum.EI_CLASS.X_64
        sec_symtab = elf.Section.MakeSectionSymTab(".symtab", which, len(sections) + 1)
        sections.append(sec_symtab)
        sec_symtab.SetData(bytearray(len(unit.symbols) * elf.Symbol.SIZE[which]))
        seg_pseudo.sections.append(sec_symtab)
        # TODO: this is not quite right
        sec_symtab.sh_info = len(unit.symbols)

        sym_names = bytearray()
        sym_names += b"\0"
        for sym in unit.symbols:
            if sym.name:
                sym.st_name = len(sym_names)
                sym_names += (bytes(sym.name, "utf-8") + b"\0")
            else:
                sym.st_name = 0
        sec_symstrtab = elf.Section.MakeSectionStrTab(".strtab")
        sections.append(sec_symstrtab)
        sec_symstrtab.SetData(sym_names)
        seg_pseudo.sections.append(sec_symstrtab)

    sec_shstrtab = elf.Section.MakeSectionStrTab(".shstrtab")
    sections.append(sec_shstrtab)
    sec_shstrtab.SetData(elf.MakeSecStrTabContents(sections))
    seg_pseudo.sections.append(sec_shstrtab)

    exe = elf.Executable.MakeExecutableA64(0x400000, sections, segments)
    exe.update_vaddrs_and_offset()

    if False:
        for sym in unit.symbols:
            print("@@@SYMS", sym)
        for rel in unit.relocations:
            print("@@@REL", rel)

    for sym in unit.symbols:
        if sym.section:
            assert sym.section.sh_addr > 0
            assert sym.st_value != elf.TO_BE_FILLED_IN_LATER
            sym.st_value += sym.section.sh_addr
            sym.st_shndx = sym.section.index

    for rel in unit.relocations:
        _ApplyRelocation(rel)

    if create_sym_tab:
        # we only put dummiess in the symtable above - do it for real now
        sec_symtab.data = bytearray()
        for sym in unit.symbols:
            sec_symtab.AddData(sym.pack(which))

    sym_entry = unit.global_symbol_map["_start"]
    assert sym_entry and not sym_entry.is_undefined()
    entry_addr = sym_entry.st_value
    print(f"PATCH ENTRY: {entry_addr:x}")
    exe.ehdr.e_entry = entry_addr
    return exe

File: CpuA64/assembler.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA64/assembler.h"
#include "CpuA64/opcode_gen.h"
#include "CpuA64/symbolic.h"
#include "Util/parse.h"

#include <map>

namespace cwerg::a64 {
using namespace cwerg::elf;

// +-prefix converts an enum the underlying type
template <typename T>
constexpr auto operator+(T e) noexcept
    -> std::enable_if_t<std::is_enum<T>::value, std::underlying_type_t<T>> {
  return static_cast<std::underlying_type_t<T>>(e);
}

std::string_view padding_zero("\0", 1);
std::string_view padding_nop("\x1f\x20\x03\xd5", 4);

void AddIns(A64Unit* unit, Ins* ins) {
  if (ins->has_reloc()) {
    const elf::Symbol<uint64_t>* sym =
        unit->FindOrAddSymbol(ins->reloc_symbol, ins->is_local_sym);
    unit->AddReloc(+ins->reloc_kind, unit->sec_text, sym,
                   ins->operands[ins->reloc_pos]);
    ins->clear_reloc();
  }
  uint32_t data = Assemble(*ins);
  unit->sec_text->AddData({(const char*)&data, 4});
}

std::array<std::string_view, 6> kStartupCode = {"ldr_x_imm x0 sp 0",    //
                                                "add_x_imm x1 sp 8",    //
                                                "bl expr:call26:main",  //
                                                "movz_x_imm x8 0x5d",   //
                                                "svc 0",                //
                                                "brk 1"};

void AddStartupCode(A64Unit* unit) {
  unit->FunStart("_start", 16, padding_nop);
  std::vector<std::string_view> token;
  for (const auto line : kStartupCode) {
    token.clear();
    if (!ParseLineWithStrings(line, false, &token)) {
      ASSERT(false, "bad internal code template " << line);
    }
    Ins ins;
    if (!InsFromSymbolized(token, &ins)) {
      ASSERT(false, "internal parse error " << token[0]);
    }
    AddIns(unit, &ins);
  }
  unit->FunEnd();
}

bool HandleDirective(A64Unit* unit,
                     const std::vector<std::string_view>& token) {
  const auto mnemonic = token[0];
  if (mnemonic == ".fun") {
    ASSERT(token.size() == 3, "");
    unit->FunStart(token[1], ParseInt<uint32_t>(token[2]).value(), padding_nop);
  } else if (mnemonic == ".endfun") {
    unit->FunEnd();
  } else if (mnemonic == ".mem") {
    ASSERT(token.size() == 4, "");
    unit->MemStart(token[1], ParseInt<uint32_t>(token[2]).value(), token[3],
                   padding_zero, false);
  } else if (mnemonic == ".localmem") {
    ASSERT(token.size() == 4, "");
    unit->MemStart(token[1], ParseInt<uint32_t>(token[2]).value(), token[3],
                   padding_zero, true);
  } else if (mnemonic == ".endmem") {
    unit->MemEnd();
  } else if (mnemonic == ".data") {
    ASSERT(token.size() == 3, "");
    size_t len = token[2].size();
    if (token[2][0] != '"' || token[2][len - 1] != '"' || len <= 2) {
      std::cerr << "malformed .data directive\n";
      return false;
    }
    char buffer[1024];
    len = EscapedStringToBytes({token[2].data() + 1, len - 2}, buffer);
    if (len == 0) {
      std::cerr << "malformed .data directive cannot parse payload\n";
      return false;
    }
    unit->AddData(ParseInt<uint32_t>(token[1]).value(), buffer, len);
  } else if (mnemonic == ".addr.mem") {
    ASSERT(token.size() == 4, "");
    unit->AddMemAddr(ParseInt<uint32_t>(token[1]).value(),
                     +RELOC_TYPE_ARM::ABS32, token[2],
                     ParseInt<uint32_t>(token[3]).value());
  } else if (mnemonic == ".addr.fun") {
    ASSERT(token.size() == 3, "");
    unit->AddFunAddr(ParseInt<uint32_t>(token[1]).value(),
                     +RELOC_TYPE_ARM::ABS32, token[2]);
  } else if (mnemonic == ".addr.bbl") {
    ASSERT(token.size() == 3, "");
    unit->AddBblAddr(ParseInt<uint32_t>(token[1]).value(),
                     +RELOC_TYPE_ARM::ABS32, token[2]);
  } else if (mnemonic == ".bbl") {
    ASSERT(token.size() == 3, "");
    unit->AddLabel(token[1], ParseInt<uint32_t>(token[2]).value(), padding_nop);
  } else {
    std::cerr << "unknown directive " << mnemonic << "\n";
    return false;
  }
  return true;
}

bool HandleOpcode(A64Unit* unit, const std::vector<std::string_view>& token) {
  Ins ins;
  if (!InsFromSymbolized(token, &ins)) {
    std::cerr << "parse error " << token[0] << "\n";
    return false;
  }
  AddIns(unit, &ins);
  return true;
}

bool UnitParse(std::istream* input, bool add_startup_code, A64Unit* unit) {
  A64Unit out;
  int line_num = 0;
  std::vector<std::string_view> token;
  for (std::string line; getline(*input, line);) {
    ++line_num;
    token.clear();
    if (!ParseLineWithStrings(line.c_str(), false, &token)) {
      std::cerr << "Error while parsing line " << line_num << "\n" << line;
      return false;
    }

    // std::cerr << "@@reading: " << line << "\n";
    if (token.empty()) continue;
    if (token[0][0] == '#') continue;
    if (token.back()[0] == '#') token.pop_back();  // get rid of comment
    // Directives
    if (token[0][0] == '.') {
      if (!HandleDirective(unit, token)) return false;
    } else {
      if (!HandleOpcode(unit, token)) return false;
    }
  }
  unit->AddLinkerDefs();
  if (add_startup_code) {
    AddStartupCode(unit);
  }
  return true;
}

int64_t BranchOffset(const Reloc<uint64_t>& rel, int64_t sym_val) {
  return (sym_val - (rel.section->shdr.sh_addr + rel.rel.r_offset)) >> 2;
}

int64_t AdrpOffset(const Reloc<uint64_t>& rel, int64_t sym_val) {
  return (sym_val >> 12ULL) -
         ((rel.section->shdr.sh_addr + rel.rel.r_offset) >> 12ULL);
}

void ApplyInsReloc(uint32_t* patch_addr, unsigned pos, uint32_t val) {
  *patch_addr = a64::Patch(*patch_addr, pos, val);
}

void ApplyRelocation(const Reloc<uint64_t>& rel) {
  void* patch_addr = (char*)rel.section->data->data() + rel.rel.r_offset;
  const int64_t sym_val = rel.symbol->sym.st_value + rel.rel.r_addend;

  switch (RELOC_TYPE_AARCH64(rel.rel.r_type)) {
    case RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21:
      ApplyInsReloc((uint32_t*)patch_addr, 1, AdrpOffset(rel, sym_val));
      return;
    case RELOC_TYPE_AARCH64::ADD_ABS_LO12_NC:
      ApplyInsReloc((uint32_t*)patch_addr, 2, sym_val & 0xfff);
      return;
    case RELOC_TYPE_AARCH64::CONDBR19:
      ApplyInsReloc((uint32_t*)patch_addr, 0, BranchOffset(rel, sym_val));
      return;
    case RELOC_TYPE_AARCH64::JUMP26:
      ApplyInsReloc((uint32_t*)patch_addr, 0, BranchOffset(rel, sym_val));
      return;
    case RELOC_TYPE_AARCH64::CALL26:
      ApplyInsReloc((uint32_t*)patch_addr, 0, BranchOffset(rel, sym_val));
      return;
    case RELOC_TYPE_AARCH64::ABS32:
      ASSERT(sym_val >> 32ULL == 0, "");
      *(uint32_t*)patch_addr = sym_val;
      return;
    case RELOC_TYPE_AARCH64::ABS64:
      *(uint64_t*)patch_addr = sym_val;
      return;
    default:
      ASSERT(false, "unknown relocation type " << rel.rel.r_type);
      return;
  }
}

Executable<uint64_t> MakeExe(A64Unit* unit, bool create_sym_tab) {
  std::vector<Section<uint64_t>*> sections;
  std::vector<Segment<uint64_t>*> segments;

  {
    auto seg_exe = new Segment<uint64_t>();
    seg_exe->InitExe(65536);
    segments.push_back(seg_exe);

    auto sec_null = new Section<uint64_t>();
    sec_null->InitNull();
    sections.push_back(sec_null);
    seg_exe->sections.push_back(sec_null);

    ASSERT(unit->sec_text->data->size() > 0, "");
    sections.push_back(unit->sec_text);
    seg_exe->sections.push_back(unit->sec_text);
  }

  if (unit->sec_rodata->data->size() > 0) {
    auto seg_ro = new Segment<uint64_t>();
    seg_ro->InitRO(65536);
    segments.push_back(seg_ro);

    sections.push_back(unit->sec_rodata);
    seg_ro->sections.push_back(unit->sec_rodata);
  }

  Segment<uint64_t>* seg_rw = nullptr;

  if (unit->sec_data->data->size() + unit->sec_bss->data->size() > 0) {
    seg_rw = new Segment<uint64_t>();
    seg_rw->InitRW(65536);
    segments.push_back(seg_rw);
  }

  if (unit->sec_data->data->size() > 0) {
    sections.push_back(unit->sec_data);
    seg_rw->sections.push_back(unit->sec_data);
  }

  if (unit->sec_bss->data->size() > 0) {
    sections.push_back(unit->sec_bss);
    seg_rw->sections.push_back(unit->sec_bss);
  }

  Section<uint64_t>* sec_symtab = nullptr;
  {
    auto seg_pseudo = new Segment<uint64_t>();
    seg_pseudo->InitPseudo();
    segments.push_back(seg_pseudo);
    //
    if (create_sym_tab) {
      sec_symtab = new Section<uint64_t>();
      sec_symtab->InitSymTab(".symtab", sizeof(Elf_Sym<uint64_t>),
                             sections.size() + 1);
      sec_symtab->shdr.sh_info = unit->symbols.size();
      std::string dummy(sizeof(Elf_Sym<uint64_t>) * unit->symbols.size(), '\0');
      auto* sym_data = new Chunk(dummy, false);
      sec_symtab->SetData(sym_data);
      sections.push_back(sec_symtab);
      seg_pseudo->sections.push_back(sec_symtab);

      auto* names = new Chunk(padding_zero, false);
      for (auto& sym : unit->symbols) {
        if (!sym->name.empty()) {
          sym->sym.st_name = names->size();
          names->AddData(sym->name);
          names->AddData(padding_zero);
        } else {
          sym->sym.st_name = 0;
        }
      }
      auto* sec_strtab = new Section<uint64_t>();
      sec_strtab->InitStrTab(".strtab");
      sec_strtab->SetData(names);
      sections.push_back(sec_strtab);
      seg_pseudo->sections.push_back(sec_strtab);
    }
    //
    auto sec_shstrtab = new Section<uint64_t>();
    sec_shstrtab->InitStrTab(".shstrtab");
    sections.push_back(sec_shstrtab);
    sec_shstrtab->SetData(MakeShStrTabContents(sections));
    seg_pseudo->sections.push_back(sec_shstrtab);
  }

  Executable<uint64_t> exe = MakeExecutableA64(0x400000, sections, segments);
  exe.ehdr.e_shoff = exe.UpdateVaddrsAndOffsets(
      CombinedElfHeaderSize<uint64_t>(exe.segments), exe.start_vaddr);
  exe.ehdr.e_phoff = sizeof(exe.ident) + sizeof(exe.ehdr);

  for (auto& sym : unit->symbols) {
    ASSERT(sym->sym.st_value != ~0, "undefined symbol " << sym->name);
    if (sym->section != nullptr) {
      ASSERT(sym->section->shdr.sh_addr != ~0,
             sym->name << "has bad sec " << *sym->section);
      sym->sym.st_value += sym->section->shdr.sh_addr;
      sym->sym.st_shndx = sym->section->index;
    }
  }

  for (auto& rel : unit->relocations) {
    ApplyRelocation(rel);
  }

  if (create_sym_tab) {
    ASSERT(sec_symtab != nullptr, "");

    auto* sym_data = new Chunk("", false);
    for (auto& sym : unit->symbols) {
      sym_data->AddData({(const char*)&sym->sym, sizeof(sym->sym)});
    }
    ASSERT(sec_symtab->shdr.sh_size == sym_data->size(), "");
    sec_symtab->SetData(sym_data);
  }

  auto* entry = unit->global_symbol_map["_start"];
  ASSERT(entry != nullptr, "_start is not defined");
  exe.ehdr.e_entry = entry->sym.st_value;
  return exe;
}

}  // namespace cwerg::a64
